File: 프로그래머스/2/84512. 모음 사전/모음 사전.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

vector<string> v;
string alpha = "AEIOU";


void DFS(string current_word, int max_length) {

    if (!current_word.empty()) {
        v.push_back(current_word);
    }

    if (current_word.length() < max_length) {
        for (char word : alpha) {
            DFS(current_word + word, max_length);
        }
    }
}


int solution(string word) {
    int answer = 0;
    
    DFS("", 5);
    sort(v.begin(), v.end());
    
    for(int i = 0; i < v.size(); i++){
        if(v[i] == word)
            return i + 1;
    }
    
    
    return answer;
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.java
import java.util.*;

public class Solution {

    static List<String> list = new ArrayList<>();
    static String str = "AEIOU";
    
    public static void DFS(String now, int cnt) {
        if (!now.isEmpty()) {
            list.add(now);
        }

        if (now.length() < cnt) {
            for (int i = 0; i < str.length(); i++) {
                DFS(now + str.charAt(i), cnt);
            }
        }
    }

    public static int solution(String word) {
        DFS("", 5);
        //Collections.sort(list);

        for (int i = 0; i < list.size(); i++) {
            if (list.get(i).equals(word)) {
                return i + 1;
            }
        }

        return 0;
    }
}
